narwhal/types/src/primary.rs

```diff
@@ -767,11 +767,28 @@ impl Affiliated for Certificate {
 
 #[derive(Clone, Debug, Serialize, Deserialize)]
 pub enum PrimaryMessage {
-    Header(Header),
-    Vote(Vote),
     Certificate(Certificate),
 }
 
+/// Used by the primary to request a vote from other primaries on newly produced headers.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct RequestVoteRequest {
+    pub header: Header,
+
+    // Optional parent certificates provided by the requester, in case this primary doesn't yet
+    // have them and requires them in order to offer a vote.
+    pub parents: Vec<Certificate>,
+}
+
+/// Used by the primary to reply to RequestVoteRequest.
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct RequestVoteResponse {
```

Comment:
Do we also want to add information to the response to indicate why a vote wasn't provided? I suppose this could be provided via an error but it would need to be done in a machine actionable way I would think. For example:

- No vote provided because the round you're asking me to vote for is too old
- Maybe you already voted for something for this round/proposer (did a crash cause some issue)
- Wrong epoch
- etc

And then in the case of some of those errors it may be beneficial to return in-band information that could help the proposer. If we're soliciting votes for a round that is too old, then maybe the other side can return to us a set of certs from the previous round, proving that the actual round is indeed X and not Y (where Y < X)

Reply:
This could go a long way towards also helping the recovery phase that @laura-makdah is worried about loosing.

Reply:
Most of the work for this is actually already done. We return a DagResult from the internal vote request handler function, which has specific DagError enums for all the cases you listed. If at some point anemo supports structured error info in responses we could send the DagError directly back to the requester. (I could file another FR?)

For the moment, instead it gets converted to a string error message, and then the only machine-actionable element is that the errors get grouped into either a BadRequest status code (for unrecoverable errors) or Unknown status code (for retriable).

Feeding wrong round, epoch, etc, back to the vote requester would be nice-to-have, but I don't think it'll substantially affect the liveness or responsiveness of the system. If node A's vote request is rejected by node B because it's too old, then node A will soon be receiving a vote request from node B (and probably other nodes too) with current round and parents. So node A will catch up eventually regardless.